File: python/logger_util.py
#!/usr/bin/env python3
"""
Logging utility for RPS system
Logs all requests, responses, and saves JSON exports
"""

import os
import json
import logging
from datetime import datetime
from pathlib import Path

# Create logs directory
LOGS_DIR = Path(__file__).parent.parent / 'logs'
EXPORTS_DIR = LOGS_DIR / 'exports'
LOGS_DIR.mkdir(exist_ok=True)
EXPORTS_DIR.mkdir(exist_ok=True)

# Setup file handlers with rotation
def setup_logger(name='rps_api'):
    """Setup logger with file and console handlers"""
    logger = logging.getLogger(name)
    logger.setLevel(logging.DEBUG)
    
    # Remove existing handlers
    logger.handlers.clear()
    
    # File handler - daily log files
    today = datetime.now().strftime('%Y-%m-%d')
    log_file = LOGS_DIR / f'python_api_{today}.log'
    
    file_handler = logging.FileHandler(log_file, encoding='utf-8')
    file_handler.setLevel(logging.DEBUG)
    
    # Console handler
    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.INFO)
    
    # Formatter
    formatter = logging.Formatter(
        '%(asctime)s | %(levelname)-8s | %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )
    file_handler.setFormatter(formatter)
    console_handler.setFormatter(formatter)
    
    logger.addHandler(file_handler)
    logger.addHandler(console_handler)
    
    return logger

# Global logger instance
logger = setup_logger()

def log_request(endpoint, data, client_ip='unknown'):
    """Log incoming request"""
    logger.info(f"📥 REQUEST | {endpoint} | IP: {client_ip}")
    logger.debug(f"Request data keys: {list(data.keys()) if isinstance(data, dict) else type(data)}")

def log_response(endpoint, status, duration_ms, error=None):
    """Log response"""
    if error:
        logger.error(f"❌ RESPONSE | {endpoint} | {status} | {duration_ms:.0f}ms | Error: {error}")
    else:
        logger.info(f"✅ RESPONSE | {endpoint} | {status} | {duration_ms:.0f}ms")

def log_timing(label, duration_ms):
    """Log timing information"""
    logger.info(f"⏱️  TIMING | {label}: {duration_ms:.2f}ms")

def save_export_json(rps_data, meta, filename_prefix='export'):
    """Save JSON data for export history"""
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    course_code = meta.get('kode', 'unknown').replace('/', '_')
    
    filename = f"{filename_prefix}_{course_code}_{timestamp}.json"
    filepath = EXPORTS_DIR / filename
    
    export_data = {
        'timestamp': datetime.now().isoformat(),
        'meta': meta,
        'rpsData': rps_data
    }
    
    try:
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(export_data, f, ensure_ascii=False, indent=2)
        logger.info(f"💾 SAVED | Export JSON: {filename}")
        return str(filepath)
    except Exception as e:
        logger.error(f"❌ SAVE ERROR | Failed to save export JSON: {e}")
        return None

def log_performance_metrics(metrics):
    """Log performance metrics"""
    logger.info("📊 PERFORMANCE METRICS:")
    for key, value in metrics.items():
        if isinstance(value, (int, float)):
            logger.info(f"   {key}: {value:.2f}ms" if value > 100 else f"   {key}: {value:.2f}s")
        else:
            logger.info(f"   {key}: {value}")

def cleanup_old_logs(days_to_keep=7):
    """Cleanup log files older than specified days"""
    import time
    cutoff_time = time.time() - (days_to_keep * 86400)
    
    deleted_count = 0
    for log_file in LOGS_DIR.glob('*.log'):
        if log_file.stat().st_mtime < cutoff_time:
            try:
                log_file.unlink()
                deleted_count += 1
            except:
                pass
    
    if deleted_count > 0:
        logger.info(f"🧹 CLEANUP | Deleted {deleted_count} old log files")

# Run cleanup on import
cleanup_old_logs()

File: java/src/main/java/com/example/openaitest/LoggerUtil.java
package com.example.openaitest;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Java port of logger_util.py
 * Logs requests, responses, timings and saves JSON exports.
 */
@Component
public class LoggerUtil {

    private static final Logger log = LoggerFactory.getLogger(LoggerUtil.class);

    // Mirrors LOGS_DIR = Path(__file__).parent.parent / 'logs'
    // Python script is at /root/AI/python/ → parent.parent = /root/ → logs = /root/logs/
    // Java project is at /root/AI/openai/ → parent.parent = /root/ → logs = /root/logs/
    private static final Path LOGS_DIR = Path.of(System.getProperty("user.dir"))
            .resolve("../../logs").normalize();

    private static final Path EXPORTS_DIR = LOGS_DIR.resolve("exports");

    static {
        try {
            Files.createDirectories(EXPORTS_DIR);
        } catch (IOException e) {
            System.err.println("⚠️ Could not create logs/exports directories: " + e.getMessage());
        }
        cleanupOldLogs(7);
    }

    // ----------------------------------------------------------------- //
    //  log_request
    // ----------------------------------------------------------------- //
    public static void logRequest(String endpoint, Map<String, Object> data, String clientIp) {
        log.info("📥 REQUEST | {} | IP: {}", endpoint, clientIp);
        log.debug("Request data keys: {}", data != null ? data.keySet() : "null");
    }

    public static void logRequest(String endpoint, Map<String, Object> data) {
        logRequest(endpoint, data, "unknown");
    }

    // ----------------------------------------------------------------- //
    //  log_response
    // ----------------------------------------------------------------- //
    public static void logResponse(String endpoint, int status, double durationMs, String error) {
        if (error != null && !error.isBlank()) {
            log.error("❌ RESPONSE | {} | {} | {}ms | Error: {}", endpoint, status,
                    String.format("%.0f", durationMs), error);
        } else {
            log.info("✅ RESPONSE | {} | {} | {}ms", endpoint, status,
                    String.format("%.0f", durationMs));
        }
    }

    public static void logResponse(String endpoint, int status, double durationMs) {
        logResponse(endpoint, status, durationMs, null);
    }

    // ----------------------------------------------------------------- //
    //  log_timing
    // ----------------------------------------------------------------- //
    public static void logTiming(String label, double durationMs) {
        log.info("⏱️  TIMING | {}: {}ms", label, String.format("%.2f", durationMs));
    }

    // ----------------------------------------------------------------- //
    //  save_export_json
    // ----------------------------------------------------------------- //
    public static String saveExportJson(Object rpsData, Map<String, Object> meta, String filenamePrefix) {
        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
        String courseCode = String.valueOf(meta.getOrDefault("kode", "unknown")).replace("/", "_");
        String filename = filenamePrefix + "_" + courseCode + "_" + timestamp + ".json";
        Path filepath = EXPORTS_DIR.resolve(filename);

        Map<String, Object> exportData = new LinkedHashMap<>();
        exportData.put("timestamp", LocalDateTime.now().toString());
        exportData.put("meta", meta);
        exportData.put("rpsData", rpsData);

        try {
            ObjectMapper mapper = new ObjectMapper();
            mapper.writerWithDefaultPrettyPrinter().writeValue(filepath.toFile(), exportData);
            log.info("💾 SAVED | Export JSON: {}", filename);
            return filepath.toString();
        } catch (Exception e) {
            log.error("❌ SAVE ERROR | Failed to save export JSON: {}", e.getMessage());
            return null;
        }
    }

    public static String saveExportJson(Object rpsData, Map<String, Object> meta) {
        return saveExportJson(rpsData, meta, "export");
    }

    // ----------------------------------------------------------------- //
    //  log_performance_metrics
    // ----------------------------------------------------------------- //
    public static void logPerformanceMetrics(Map<String, Object> metrics) {
        log.info("📊 PERFORMANCE METRICS:");
        for (Map.Entry<String, Object> entry : metrics.entrySet()) {
            Object v = entry.getValue();
            if (v instanceof Number) {
                double d = ((Number) v).doubleValue();
                log.info("   {}: {}", entry.getKey(),
                        d > 100 ? String.format("%.2fms", d) : String.format("%.2fs", d));
            } else {
                log.info("   {}: {}", entry.getKey(), v);
            }
        }
    }

    // ----------------------------------------------------------------- //
    //  cleanup_old_logs
    // ----------------------------------------------------------------- //
    public static void cleanupOldLogs(int daysToKeep) {
        try {
            if (!Files.exists(LOGS_DIR)) return;
            long cutoffMs = System.currentTimeMillis() - (long) daysToKeep * 86_400_000L;
            long[] deleted = {0};
            Files.list(LOGS_DIR)
                    .filter(p -> p.toString().endsWith(".log"))
                    .filter(p -> {
                        try { return Files.getLastModifiedTime(p).toMillis() < cutoffMs; }
                        catch (IOException e) { return false; }
                    })
                    .forEach(p -> {
                        try { Files.delete(p); deleted[0]++; }
                        catch (IOException ignored) {}
                    });
            if (deleted[0] > 0) {
                log.info("🧹 CLEANUP | Deleted {} old log files", deleted[0]);
            }
        } catch (IOException ignored) {}
    }
}
